### PR TITLE
Add an example application that uses the pseudoconsole APIs

### PR DESCRIPTION
**Pull request by bitcrazed**
Sample app to illustrate how to use the new ConPTY APIs & infrastructure.

### CODE REVIEW

File: samples/ConPTY/EchoCon/EchoCon/EchoCon.vcxproj [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+    <Import Project="$(UserRootDir)\Microsoft.Cpp.$(Platform).user.props" Condition="exists('$(UserRootDir)\Microsoft.Cpp.$(Platform).user.props')" Label="LocalAppDataPlatform" />
+  </ImportGroup>
+  <ImportGroup Label="PropertySheets" Condition="'$(Configuration)|$(Platform)'=='Release|Win32'">
+    <Import Project="$(UserRootDir)\Microsoft.Cpp.$(Platform).user.props" Condition="exists('$(UserRootDir)\Microsoft.Cpp.$(Platform).user.props')" Label="LocalAppDataPlatform" />
+  </ImportGroup>
+  <ImportGroup Label="PropertySheets" Condition="'$(Configuration)|$(Platform)'=='Debug|x64'">
+    <Import Project="$(UserRootDir)\Microsoft.Cpp.$(Platform).user.props" Condition="exists('$(UserRootDir)\Microsoft.Cpp.$(Platform).user.props')" Label="LocalAppDataPlatform" />
+  </ImportGroup>
+  <ImportGroup Label="PropertySheets" Condition="'$(Configuration)|$(Platform)'=='Release|x64'">
+    <Import Project="$(UserRootDir)\Microsoft.Cpp.$(Platform).user.props" Condition="exists('$(UserRootDir)\Microsoft.Cpp.$(Platform).user.props')" Label="LocalAppDataPlatform" />
```
Comment (DHowett-MSFT) [inline samples/ConPTY/EchoCon/EchoCon/EchoCon.vcxproj]: lot of duplication here; can you eliminate common subexpressions?

Comment (bitcrazed) [reply samples/ConPTY/EchoCon/EchoCon/EchoCon.vcxproj]: Let's leave VS to play around with the VCXPROJ.

File: samples/ConPTY/EchoCon/EchoCon/EchoCon.cpp [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+#include <process.h>
+#include <wchar.h>
+
+void __cdecl PipeListener(LPVOID);
+HRESULT InitializeStartupInfoAttachedToConPTY(STARTUPINFOEX*, HPCON);
+COORD GetConsoleSize(HANDLE);
+
+int main()
+{
+    TCHAR szCommand[] = L"ping localhost";
```
Comment (DHowett-MSFT) [inline samples/ConPTY/EchoCon/EchoCon/EchoCon.cpp]: you used `TCAHR` and `L` together: big no-no! either standardize on `W` APIs everywhere and eat the bullet with `wchar_t`, or don't commit this :)

File: samples/ConPTY/EchoCon/EchoCon/EchoCon.cpp [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+    HRESULT hr = S_OK;
+    HANDLE hPipeIn = 0;
```
Comment (DHowett-MSFT) [inline samples/ConPTY/EchoCon/EchoCon/EchoCon.cpp]: prefer `INVALID_HANDLE_VALUE`

File: samples/ConPTY/EchoCon/EchoCon/EchoCon.cpp [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+    HANDLE hPipeOut = 0;
+    HANDLE hPipePTYIn = 0;
+    HANDLE hPipePTYOut = 0;
+    HPCON hPC = 0;
+
+    PROCESS_INFORMATION piClient{};
```
Comment (DHowett-MSFT) [inline samples/ConPTY/EchoCon/EchoCon/EchoCon.cpp]: please standardize on one type of initialization; this is a hodge-podge of C++11 unified and old C style.

File: samples/ConPTY/EchoCon/EchoCon/EchoCon.cpp [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+            HANDLE hPipeListenerThread = (HANDLE)_beginthread(
+                PipeListener,           // Thread function
+                0,                      // Default stack
+                hPipeIn);               // Thread arg
+        
+            // Create ConPTY
+            hr = CreatePseudoConsole(consoleSize, hPipePTYIn, hPipePTYOut, 0, &hPC);
+
+            // Initialize the necessary startup info struct        
+            if (S_OK == InitializeStartupInfoAttachedToConPTY(&startupInfo, hPC))
```
Comment (DHowett-MSFT) [inline samples/ConPTY/EchoCon/EchoCon/EchoCon.cpp]: please standardize on `PseudoConsole` -- the other APIs look like it.

File: samples/ConPTY/EchoCon/EchoCon/EchoCon.cpp [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+                // Cleanup attribute list
+                if (startupInfo.lpAttributeList)
+                {
+                    DeleteProcThreadAttributeList(startupInfo.lpAttributeList);
+                    free(startupInfo.lpAttributeList);
+                }
+            }
+            else
+            {
+                _tprintf_s(L"Error: Failed to initialize StartupInfo attached to ConPTY [0x%x]",
```
Comment (DHowett-MSFT) [inline samples/ConPTY/EchoCon/EchoCon/EchoCon.cpp]: another `T` + `W` mix

File: samples/ConPTY/EchoCon/EchoCon/EchoCon.cpp [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+            ? S_OK
+            : GetLastError();
+    }
+
+    return hr == S_OK;
+}
+
+void __cdecl PipeListener(LPVOID pipe)
+{
+    HANDLE hPipe = (HANDLE)(HANDLE*)pipe;
```
Comment (DHowett-MSFT) [inline samples/ConPTY/EchoCon/EchoCon/EchoCon.cpp]: double cast, don't do that. also maybe prefer `static_cast` or `reinterpret_cast`

File: samples/ConPTY/EchoCon/EchoCon/EchoCon.cpp [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+    HANDLE hConsole = GetStdHandle(STD_OUTPUT_HANDLE);
+
+    DWORD dwBytesWritten = 0;
+    const DWORD BUFF_SIZE = 512;
+    char* pszBuffer = (char*)malloc(BUFF_SIZE);
```
Comment (DHowett-MSFT) [inline samples/ConPTY/EchoCon/EchoCon/EchoCon.cpp]: this is C++! use `new char[BUFF_SIZE]` or a unique_ptr or vector or something

File: samples/ConPTY/EchoCon/EchoCon/EchoCon.cpp [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+
+            // Write received text to the Console
+            WriteFile(hConsole, pszBuffer, dwBytesRead, &dwBytesWritten, NULL);
+
+        } while (fRead && dwBytesRead >= 0);
+
+        free(pszBuffer);
+
+        // Automatically cleans-up
+        _endthread();
```
Comment (DHowett-MSFT) [inline samples/ConPTY/EchoCon/EchoCon/EchoCon.cpp]: Prefer not doing this; just return.

>You can call _endthread or _endthreadex explicitly to terminate a thread; however, _endthread or _endthreadex is called automatically when the thread returns from the routine passed as a parameter to _beginthread or _beginthreadex.

File: samples/ConPTY/EchoCon/EchoCon/EchoCon.cpp [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+        InitializeProcThreadAttributeList(NULL, 1, 0, &size);
+
+        // Allocate a thread attribute list of the correct size
+        pStartupInfo->lpAttributeList = (LPPROC_THREAD_ATTRIBUTE_LIST)malloc((size_t)size);
+
+        // Initialize thread attribute list
+        if (pStartupInfo->lpAttributeList 
+            && InitializeProcThreadAttributeList(pStartupInfo->lpAttributeList, 1, 0, &size))
+        {
+            // Set thread attribute list's Pseudo Console to the specified ConPTY
```
Comment (DHowett-MSFT) [inline samples/ConPTY/EchoCon/EchoCon/EchoCon.cpp]: this comment makes it sound like a ConPTY is a _type of_ pseudoconsole

File: samples/ConPTY/EchoCon/EchoCon/EchoCon.cpp (hunk 2 above)
Comment (bitcrazed) [reply samples/ConPTY/EchoCon/EchoCon/EchoCon.cpp]: Fair. Fixed.

File: samples/ConPTY/EchoCon/EchoCon/EchoCon.cpp (hunk 6 above)
Comment (bitcrazed) [reply samples/ConPTY/EchoCon/EchoCon/EchoCon.cpp]: Done.

File: samples/ConPTY/EchoCon/EchoCon/EchoCon.cpp [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+            // Initialize the necessary startup info struct        
+            if (S_OK == InitializeStartupInfoAttachedToPseudoConsole(&startupInfo, hPC))
+            {
+                // Launch ping to echo some text back via the pipe
+                hr = CreateProcess(
+                    NULL,                           // No module name - use Command Line
+                    szCommand,                      // Command Line
+                    NULL,                           // Process handle not inheritable
+                    NULL,                           // Thread handle not inheritable
+                    TRUE,                           // Inherit handles
```
Comment (zadjii-msft) [inline samples/ConPTY/EchoCon/EchoCon/EchoCon.cpp]: I believe it's safer to pass FALSE for this param.

I thought we fixed whatever the issue was with the client app inheriting the handles of the console application acting as a terminal, but for whatever reason node-pty was still acting weird if we passed TRUE here. 

I think it'd be safer to have this param as FALSE, so no-one else hits this gotcha, at least until we figure out exactly what's wrong with node-pty.

**Comment by zadjii-msft** [comment]
@DHowett-MSFT re: your concern on what this sample demos - 

Would you think it'd be a better sample if the command he was running was `ping -t localhost`, and after <some timeout>, he wrote Ctrl-C to the conpty to interrupt ping.exe? That would demo reading output and also sending not just input, but input that was previously impossible to send to the console with the console API.

@bitcrazed thoughts?

File: samples/ConPTY/EchoCon/EchoCon/EchoCon.cpp [hunk 12]
```diff
@@ -0,0 +1,197 @@
...
+        {
+            // Create & start thread to listen to the incoming pipe
+            // NB: Using CRT-safe _beginthread() rather than CreateThread()
+            DWORD pipeListenerThreadId{};
+            HANDLE hPipeListenerThread{ 
+                reinterpret_cast<HANDLE>(_beginthread(PipeListener, 0, hPipeIn)) };
+
+            // Create appropriately sized ConPTY
+            COORD consoleSize{ GetConsoleSize(hConsole) };
+            hr = CreatePseudoConsole(consoleSize, hPipePTYIn, hPipePTYOut, 0, &hPC);
```
[the remark this answers is not in the document]
Comment (bitcrazed) [reply samples/ConPTY/EchoCon/EchoCon/EchoCon.cpp]: Good suggestion. On it now. Confirmed with @zadjii-msft that the ConPTY pipe ends are duped into ConHost so can be released by openpty() type function

File: samples/ConPTY/EchoCon/EchoCon/EchoCon.cpp [hunk 13]
```diff
@@ -0,0 +1,197 @@
...
+        }
+
+        // Clean-up the pipes
+        if (hPipePTYOut != INVALID_HANDLE_VALUE) CloseHandle(hPipePTYOut);
+        if (hPipePTYIn != INVALID_HANDLE_VALUE) CloseHandle(hPipePTYIn);
+        if (hPipeOut != INVALID_HANDLE_VALUE) CloseHandle(hPipeOut);
+        if (hPipeIn != INVALID_HANDLE_VALUE) CloseHandle(hPipeIn);
+
+        // Restore console to its original mode.
+        hr = SetConsoleMode(hConsole, consoleMode)
```
[the remark this answers is not in the document]
Comment (bitcrazed) [reply samples/ConPTY/EchoCon/EchoCon/EchoCon.cpp]: Good to know. Thanks.

File: samples/ConPTY/EchoCon/EchoCon/EchoCon.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (bitcrazed) [reply samples/ConPTY/EchoCon/EchoCon/EchoCon.cpp]: Added a CreateConPTY() helper.

File: samples/ConPTY/EchoCon/EchoCon/EchoCon.cpp [hunk 14]
```diff
@@ -0,0 +1,197 @@
...
+    if (S_OK == hr)
+    {
+        // Create the pipes to which the ConPTY will connect
+        if (CreatePipe(&hPipePTYIn, &hPipeOut, NULL, 0) &&
+            CreatePipe(&hPipeIn, &hPipePTYOut, NULL, 0))
+        {
+            // Create & start thread to listen to the incoming pipe
+            // NB: Using CRT-safe _beginthread() rather than CreateThread()
+            DWORD pipeListenerThreadId{};
+            HANDLE hPipeListenerThread{
```
[the remark this answers is not in the document]
Comment (bitcrazed) [reply samples/ConPTY/EchoCon/EchoCon/EchoCon.cpp]: Yep - thread creation now happens naturally after CreateConPTY() call.

File: samples/ConPTY/EchoCon/EchoCon/EchoCon.cpp (hunk 11 above)
Comment (bitcrazed) [reply samples/ConPTY/EchoCon/EchoCon/EchoCon.cpp]: Thanks. Fixed.

File: samples/ConPTY/EchoCon/EchoCon/EchoCon.cpp [hunk 15]
```diff
@@ -0,0 +1,197 @@
...
+
+    DWORD dwBytesWritten{};
+    DWORD dwBytesRead{};
+    BOOL fRead{ FALSE };
+    do
+    {
+        // Read from the pipe
+        fRead = ReadFile(hPipe, pszBuffer, BUFF_SIZE, &dwBytesRead, NULL);
+
+        // Write received text to the Console
```
[the remark this answers is not in the document]
Comment (bitcrazed) [reply samples/ConPTY/EchoCon/EchoCon/EchoCon.cpp]: Good idea. 👍

File: samples/ConPTY/EchoCon/EchoCon/EchoCon.cpp [hunk 16]
```diff
@@ -1,64 +1,53 @@
...
 
 #include "stdafx.h"
 #include <Windows.h>
-#include <conio.h>
 #include <process.h>
-#include <wchar.h>
 
 // Forward declarations
-void __cdecl PipeListener(LPVOID);
+HRESULT CreateConPTY(HPCON*, HANDLE*, HANDLE*);
```
Comment (DHowett-MSFT) [inline samples/ConPTY/EchoCon/EchoCon/EchoCon.cpp]: Can you standardize on one or the other? Please?

File: samples/ConPTY/EchoCon/EchoCon/EchoCon.cpp [hunk 17]
```diff
@@ -69,18 +58,16 @@ int main()
...
                     // Allow listening thread to catch-up with final output!
                     Sleep(500);
                 }
 
                 // --- CLOSEDOWN ---
-                // Close ConPTY - this will terminate client process if running
-                ClosePseudoConsole(hPC);
 
-                // Now safe to clean-up client process info process & thread
+                // Now safe to clean-up client app's process-info & thread
```
Comment (DHowett-MSFT) [inline samples/ConPTY/EchoCon/EchoCon/EchoCon.cpp]: canonically, just `process info` (no `-`)

File: samples/ConPTY/EchoCon/EchoCon/EchoCon.cpp [hunk 18]
```diff
@@ -91,52 +78,50 @@ int main()
...
-        if (hPipeIn != INVALID_HANDLE_VALUE) CloseHandle(hPipeIn);
+            // Close ConPTY - this will terminate client process if running
+            ClosePseudoConsole(hPC);
 
-        // Restore console to its original mode.
-        hr = SetConsoleMode(hConsole, consoleMode)
-            ? S_OK
-            : GetLastError();
+            // Clean-up the pipes
+            if (INVALID_HANDLE_VALUE != hPipeOut) CloseHandle(hPipeOut);
```
Comment (DHowett-MSFT) [inline samples/ConPTY/EchoCon/EchoCon/EchoCon.cpp]: y u switch these conditionals around?

File: samples/ConPTY/EchoCon/EchoCon/EchoCon.cpp [hunk 19]
```diff
@@ -182,16 +167,26 @@ HRESULT InitializeStartupInfoAttachedToPseudoConsole(STARTUPINFOEX* pStartupInfo
...
 
-COORD GetConsoleSize(HANDLE hStdOut)
+void __cdecl PipeListener(LPVOID pipe)
 {
-    COORD size = { 0, 0 };
+    HANDLE hPipe{ pipe };
+    HANDLE hConsole{ GetStdHandle(STD_OUTPUT_HANDLE) };
+
+    const DWORD BUFF_SIZE{ 512 };
+    char szBuffer[BUFF_SIZE]{};
```
Comment (DHowett-MSFT) [inline samples/ConPTY/EchoCon/EchoCon/EchoCon.cpp]: consider this:

```
char szBuffer[512]{};
…
ReadFile(hPipe, szBuffer, std::extent<decltype(szBuffer)::value, …)
```

to kill a silly YELLING_CASE constant

File: samples/ConPTY/EchoCon/EchoCon/EchoCon.cpp (hunk 16 above)
Comment (bitcrazed) [reply samples/ConPTY/EchoCon/EchoCon/EchoCon.cpp]: The `PipeListener()` is a thread handler and has to conform to the signature expected by the CRT-safe `_beginThread()`

Comment (DHowett-MSFT) [reply samples/ConPTY/EchoCon/EchoCon/EchoCon.cpp]: That's entirely immaterial to my comment.
about "ConPTY", _Can you standardize on one or the other, please?_

File: samples/ConPTY/EchoCon/EchoCon/EchoCon.cpp (hunk 19 above)
Comment (bitcrazed) [reply samples/ConPTY/EchoCon/EchoCon/EchoCon.cpp]: Why use 9 chars when a namespaced templated expression will suffice. Doesn't add anything of significant value and just complicates things for the casual observer.

File: samples/ConPTY/EchoCon/EchoCon/EchoCon.cpp (hunk 18 above)
Comment (bitcrazed) [reply samples/ConPTY/EchoCon/EchoCon/EchoCon.cpp]: I prefer `<const> == <variable>` to protect against accidental assignment. Ol' trick I learned from Scott Myers' "[Effective C++](https://www.amazon.com/Effective-Specific-Improve-Programs-Designs/)" that's saved my butt on several occasions.

File: samples/ConPTY/EchoCon/EchoCon/EchoCon.cpp (hunk 17 above)
Comment (bitcrazed) [reply samples/ConPTY/EchoCon/EchoCon/EchoCon.cpp]: nit

File: samples/ConPTY/EchoCon/EchoCon/EchoCon.cpp [hunk 20]
```diff
@@ -0,0 +1,192 @@
...
+                }
+
+                // --- CLOSEDOWN ---
+
+                // Now safe to clean-up client app's process-info & thread
+                CloseHandle(piClient.hThread);
+                CloseHandle(piClient.hProcess);
+
+                // Cleanup attribute list
+                if (startupInfo.lpAttributeList)
```
[the remark this answers is not in the document]
Comment (bitcrazed) [reply samples/ConPTY/EchoCon/EchoCon/EchoCon.cpp]: True. Removed. Thanks.

**Pull request merged**